### PR TITLE
feat: scheduler source (3/): add member cluster scheduler source controller

### PR DESCRIPTION
**Pull request by michaelawyu**
### Description of your changes

This PR adds a scheduler source controller for processing CRP removal.

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

- [x] Integration tests

### Special notes for your reviewer
 
This PR overlaps with some of the other PRs.

### CODE REVIEW

File: pkg/scheduler/watchers/membercluster/controller.go [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+			// Capture label changes.
+			//
+			// Note that the controller runs only when label changes happen on joined clusters.
+			if newCluster.Spec.State == fleetv1beta1.ClusterStateJoin && !reflect.DeepEqual(oldCluster.Labels, newCluster.Labels) {
+				return true
+			}
+
+			// The cluster has left.
+			if oldCluster.Spec.State == fleetv1beta1.ClusterStateJoin && newCluster.Spec.State == fleetv1beta1.ClusterStateLeave {
```
Comment (zhiying-lin) [inline pkg/scheduler/watchers/membercluster/controller.go:207]: do we need to handle when cluster is left and then rejoin?

File: pkg/scheduler/watchers/membercluster/controller_integration_test.go [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+			Expect(hubClient.Get(ctx, types.NamespacedName{Name: clusterName2}, memberCluster)).To(Succeed(), "Failed to get member cluster")
+
+			// Update the status; mark the cluster as ready.
+			memberCluster.Labels = map[string]string{
+				dummyLabel: dummyLabelValue,
+			}
+			Expect(hubClient.Update(ctx, memberCluster)).To(Succeed(), "Failed to update member cluster status")
+		})
+
+		It("should enqueue CRPs (case 1b)", func() {
```
Comment (zhiying-lin) [inline pkg/scheduler/watchers/membercluster/controller_integration_test.go:106]: it's handling 1a? label has changed?

File: pkg/scheduler/watchers/membercluster/controller_integration_test.go [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+
+			// Retrieve the cluster.
+			memberCluster := &fleetv1beta1.MemberCluster{}
+			Expect(hubClient.Get(ctx, types.NamespacedName{Name: clusterName2}, memberCluster)).To(Succeed(), "Failed to get member cluster")
+
+			// Update the status; mark the cluster as ready.
+			memberCluster.Labels = map[string]string{
+				dummyLabel: dummyLabelValue,
+			}
+			Expect(hubClient.Update(ctx, memberCluster)).To(Succeed(), "Failed to update member cluster status")
```
Comment (zhiying-lin) [inline pkg/scheduler/watchers/membercluster/controller_integration_test.go:103]: wrong comments & messages?

it's updating the labels?

File: pkg/scheduler/watchers/membercluster/controller_integration_test.go [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+		})
+
+		It("should not enqueue CRPs for the left cluster", func() {
+			Eventually(noKeyEnqueuedActual, eventuallyDuration, eventuallyInterval).Should(Succeed(), "Workqueue is not empty")
+			Consistently(noKeyEnqueuedActual, consistentlyDuration, consistentlyInterval).Should(Succeed(), "Workqueue is not empty")
+		})
+
+		AfterAll(func() {
+			keyCollector.Reset()
+		})
```
[the remark this answers is not in the document]
Comment (zhiying-lin) [reply pkg/scheduler/watchers/membercluster/controller_integration_test.go:113]: i guess chen wants to make the setup/shutdown steps clear.

File: pkg/scheduler/watchers/membercluster/controller.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+		r.SchedulerWorkQueue.AddRateLimited(queue.ClusterResourcePlacementKey(crp.Name))
+	}
+
+	// The reconciliation loop completes.
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager builds a controller with Reconciler and sets it up with a controller manager.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	customPredicate := predicate.Funcs{
```
Comment (ryanzhang-oss) [inline pkg/scheduler/watchers/membercluster/controller.go:157]: This function has no logs at all. We will be in the dark if anything unexpected happens.

**Review by ryanzhang-oss — APPROVED** [review]

File: pkg/scheduler/watchers/membercluster/controller.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+			// during the downtime; in other words, notifications from this controller is not
+			// necessary.
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			// Ignore deletion events; the removal of a cluster is first signaled by the join
+			// state change (join -> leave), which is already handled separately.
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
```
[the remark this answers is not in the document]
Comment (zhiying-lin) [reply pkg/scheduler/watchers/membercluster/controller.go:174]: The genericFunc by default will return true and be processed.


// Generic implements Predicate.
func (p Funcs) Generic(e event.GenericEvent) bool {
	if p.GenericFunc != nil {
		return p.GenericFunc(e)
	}
	return true
}

**Comment by michaelawyu** [comment]
Merging this PR to unblock progress (approval acquired previously); if there's any concern, please let me know.

**Pull request merged**